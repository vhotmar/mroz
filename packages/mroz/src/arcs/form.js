import isFunction from "lodash/isFunction";
import { startSubmit, stopSubmit } from "redux-form";
import actions from "./actions.js";

export default function form(form, getErrors) {
  const fn = isFunction(form) ? form : () => form;
  const errFn = isFunction(getErrors) ? getErrors : () => undefined;

  return actions(
    config => startSubmit(fn(config)),
    (data, config) => stopSubmit(fn(config)),
    (error, config) => stopSubmit(fn(config), errFn(error, config))
  );
}
