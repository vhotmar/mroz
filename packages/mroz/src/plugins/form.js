import { reducer as formReducer } from "redux-form";
import { addReducers } from "./utils";

export default ({ reducers = {} } = {}) =>
  addReducers({ form: formReducer.plugin(reducers) });
