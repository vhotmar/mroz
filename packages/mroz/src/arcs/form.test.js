import form from "./form";
import { Observable } from "rxjs";
import asyncTest, { config } from "../epics/async.test.utils";

describe("form arc", () => {
  it("should dispatch actions", () => {
    const config = form("sample_form");

    asyncTest({
      config,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });

    asyncTest({
      work: () => Observable.throw("error"),
      config,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });
  });

  it("should map errors", () => {
    const error = { data: "error" };
    const errorFunction = jest.fn(() => error);

    asyncTest({
      work: () => Observable.throw("error"),
      config: form("sample_form", errorFunction),
      callback: res => {
        expect(res).toMatchSnapshot();
        expect(errorFunction).toHaveBeenCalledTimes(1);
        expect(errorFunction).toHaveBeenCalledWith("error", config);
      }
    });
  });

  it("should accept function as form name", () => {
    const formFunction = jest.fn(() => "sample_form");

    asyncTest({
      config: form(formFunction),
      callback: res => {
        expect(res).toMatchSnapshot();
        expect(formFunction).toHaveBeenCalledTimes(2); // request, success == 2 times
        expect(formFunction).toHaveBeenCalledWith(config);
      }
    });
  });
});
