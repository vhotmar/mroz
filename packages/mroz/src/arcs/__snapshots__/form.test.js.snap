// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form arc should accept function as form name 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
      "form": "sample_form",
    },
    "type": "@@redux-form/START_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "error": false,
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
      "form": "sample_form",
    },
    "payload": undefined,
    "type": "@@redux-form/STOP_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`form arc should dispatch actions 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
      "form": "sample_form",
    },
    "type": "@@redux-form/START_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "error": false,
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
      "form": "sample_form",
    },
    "payload": undefined,
    "type": "@@redux-form/STOP_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`form arc should dispatch actions 2`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
      "form": "sample_form",
    },
    "type": "@@redux-form/START_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "error": false,
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
      "form": "sample_form",
    },
    "payload": undefined,
    "type": "@@redux-form/STOP_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
]
`;

exports[`form arc should map errors 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
      "form": "sample_form",
    },
    "type": "@@redux-form/START_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "error": true,
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
      "form": "sample_form",
    },
    "payload": Object {
      "data": "error",
    },
    "type": "@@redux-form/STOP_SUBMIT",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
]
`;
